src/ophyd_async/core/device.py

```diff
@@ -83,12 +91,21 @@ async def connect(self, sim: bool = False, timeout: float = DEFAULT_TIMEOUT):
         timeout:
             Time to wait before failing with a TimeoutError.
         """
-        coros = {
-            name: child_device.connect(sim, timeout=timeout)
-            for name, child_device in self.children()
-        }
-        if coros:
-            await wait_for_connection(**coros)
+        previous_connect_ok = (
+            self._connect_task
+            and self._connect_task.done()
+            and not self._connect_task.exception()
+        )
+        if force_reconnect or not previous_connect_ok:
+            # Kick off a connection
+            coros = {
+                name: child_device.connect(sim, timeout=timeout)
+                for name, child_device in self.children()
+            }
+            self._connect_task = asyncio.create_task(wait_for_connection(**coros))
+        # Wait for it to complete
+        assert self._connect_task
+        await self._connect_task
```

Comment:
It so that we don't get this issue 
`"None" is not awaitable
  "None" is incompatible with protocol "Awaitable[_T_co@Awaitable]"
    "__await__" is not presentPylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)`